circuit/environment/src/traits/inject.rs

```diff
@@ -68,44 +68,21 @@ impl<C: Inject<Primitive = P>, P> Inject for Vec<C> {
 /****** Tuples ******/
 /********************/
 
-impl<C0: Inject, C1: Inject> Inject for (C0, C1) {
-    type Primitive = (C0::Primitive, C1::Primitive);
-
-    #[inline]
-    fn new(mode: Mode, value: Self::Primitive) -> Self {
-        (C0::new(mode, value.0), C1::new(mode, value.1))
-    }
-}
-
-impl<C0: Inject, C1: Inject, C2: Inject> Inject for (C0, C1, C2) {
-    type Primitive = (C0::Primitive, C1::Primitive, C2::Primitive);
-
-    #[inline]
-    fn new(mode: Mode, value: Self::Primitive) -> Self {
-        (C0::new(mode, value.0), C1::new(mode, value.1), C2::new(mode, value.2))
-    }
-}
-
-impl<C0: Inject, C1: Inject, C2: Inject, C3: Inject> Inject for (C0, C1, C2, C3) {
-    type Primitive = (C0::Primitive, C1::Primitive, C2::Primitive, C3::Primitive);
-
-    #[inline]
-    fn new(mode: Mode, value: Self::Primitive) -> Self {
-        (C0::new(mode, value.0), C1::new(mode, value.1), C2::new(mode, value.2), C3::new(mode, value.3))
+/// A helper macro to implement `Inject` for a tuple of `Inject` circuits.
+macro_rules! inject_tuple {
+    (($t0:ident, $i0:expr), $(($ty:ident, $idx:tt)),*) => {
+        impl<$t0: Inject, $($ty: Inject),*> Inject for ($t0, $($ty),*) {
+            type Primitive = ($t0::Primitive, $( $ty::Primitive ),*);
+
+            #[inline]
+            fn new(mode: Mode, value: Self::Primitive) -> Self {
+                ($t0::new(mode, value.0), $($ty::new(mode, value.$idx)),*)
```

Comment:
```suggestion
    (($t0:ident, $i0:tt), $(($ty:ident, $idx:tt)),*) => {
        impl<$t0: Inject, $($ty: Inject),*> Inject for ($t0, $($ty),*) {
            type Primitive = ($t0::Primitive, $( $ty::Primitive ),*);

            #[inline]
            fn new(mode: Mode, value: Self::Primitive) -> Self {
                ($t0::new(mode, value.$i0), $($ty::new(mode, value.$idx)),*)
```
I'd suggest to either do this (i.e. align and use `$i0`) or to change the first param from `($t0:ident, $i0:tt)` to `(C0, 0)` in order to be consistent in its use

Reply:
But `$t0::Primitive` is always the first element in the tuple whether `$i0` is `0` or not, so I'd prefer to stick to it, which is consistent with `eject_tuple`.

Reply:
I see how this is aligned with `eject_tuple` - `$i0` unused in both; this is a good opportunity to fix it in both places, by at least changing `$i0:expr` to `0`

Reply:
Good point! Changed accordingly.